### PR TITLE
Add cmds/core/netstat

### PR DESCRIPTION
**Pull request by ChriMarMe**
Snapshot of status quo netstat implementation in u-root compared to [net-tools/netstat](https://github.com/ecki/net-tools/blob/master/netstat.c)

netstat has five main capabilities:

- List socket states
- List route table (flags: -r, -r4,-r6,-r46)
- List multicast group membership (flags: -g, -g4, -g6, -g46)
- Print interface information(s) (flags: -i, -I=)
- Print network statistics (flags: -s, -s4, -s6)

Status of implementation:
- [x] printing of sockets (tcp, udp, raw for IPv4/IPv6, unix) (netstat -4, netstat -6, netstat -46, netstat -r, netstat -x)
- [x] printing of routing table (IPv4, IPv6) (netstat -r, netstat -r4, netstat -r6, netstat -r46)
- [x] printing of statistics (IPv4, IPv6) (netstat -s, netstat -s4, netstat -s6)
- [x] printing of multicast group memberships (IPv4, IPv6) (netstat -g, netstat -g4, netstat -g6, netstat -g46)
- [x] printing of interfaces (netstat -i)
- [x] printing of single interface (netstat -I="ifacename")
- [x] flag `--tcp`, `--udp` (for IPv4, IPv6)
- [x] flag `--all` (display all connections, default: connected)
- [ ] flag `--verbose`, `-v` (Provides errors for unsupported address families of the kernel)
>Does not add value to output, core functionality maps to kernel core functionality
- [x] `--wide` (don't truncate IP addresses)
- [x] `--extend`(display other/more information) 
- [x] `--programs` (display PID/Program name for sockets)
- [x] `--timers` (display timers)
- [x] `--listening` (display listening server sockets)
- [x] `--continuous` (continuous listing)
- [x] `--cache` (display routing cache instead of FIB)
- [ ] `--numeric` (don't resolve names)
- [x] `--numeric-host` (don't resolve host names)
- [x] `--numeric-user` (don't resolve user names)
- [ ] `--numeric-port` (don't resolve port names)
- [ ] `--symbolic` (resolve hardware names)
- [x] support socket type tcp
- [x] support socket type udp 
- [x] support socket type `raw`
- [x] support socket type `unix`

### CODE REVIEW

**Comment by jensdrenhaus** [comment]
**What socket types and address families should be supported?** 

Socket types:
- raw
- unix
- tcp
- udp
- ax25
- sctp (Stream Control Transmission Protocol)
- ipx
- netrom

Address families:
- IPv4
- IPv6
- ax25 (Amateur radio AX.25 protocol)
- netrom (AMPR NET/ROM)
- ipx (Novell IPX)
- ddp(Appletalk DDP)
- x25 (ITU-T X.25 / ISO-8208 protocol)

@ChriMarMe evaluated the socket type usage as follows:
**raw**
raw sockets allow the manipulation of payload and header for the given address family in layer 2 (data-link-layer) or layer 3 (network-layer)
https://stackoverflow.com/questions/14774668/what-is-raw-socket-in-socket-programming
This is a common, broadly used socket type.

**unix**
unix sockets provide a mechanism for inter-process communication defined in POSIX standard and implemented in the linux kernel.
This is a common, broadly used socket type.

**tcp/udp**
sockets of type tcp/udp usually used on top of IPv4/IPv6 address families.
This is a common, broadly used socket type

**ax25**
Broadly used in packet radio application 
IIuc: Drop-in replacement for ethernet for a special use case (digital radio stuff) There is no usage in u-root so far. 
sctp
The “Stream Control Transmission Protocol” is a transport layer protocol and provides UDP-like connections with TCP reliability build-in.
There is no usage in u-root so far. 

**ipx**
Proprietary protocol. Works like IP/UDP. Popular in the past, not so much anymore. 
Does not see usage in u-root

**netrom**
netrom is a protocol used extensively by radio amateurs. The Linux netrom protocol family permits access to these protocols via the standard networking socket metaphor. 
Does not see use in u-root

----------------------------------------------------------------------------------------------

**Proposal:**
Processing of the RAW and UNIX socket type data is a viable thing to do. Others do not see any usage in u-root nor vendored dependencies. We suggest implementing RAW and UNIX socket type data processing and skipping the rest. Same goes for the address families linked to the skipped socket types. (ax25,x25,ddp,ipx, netrom)

**Comment by jensdrenhaus** [comment]
cc @10000TB @andrewsun2898

File: pkg/netstat/route.go
Comment (rminnich) [inline pkg/netstat/route.go:38]: I have to wonder: could this be an array: 
type flags []struct {
flag uint32 
n string}{
RFGATEWAY, "G"
...
}
for _, f := range flags {
if flag& f.flag {
s.WriteString(f.n)
}

**Review by rminnich — COMMENTED** [review]
this is pretty amazing. See my one comment. You might put the info on what is done, and what is not done, in doc.go

I would be ready to approve as it is now if you want.

File: pkg/netstat/route.go
Comment (ChriMarMe) [reply pkg/netstat/route.go]: Thanks for the suggestion. Much better way to do things. Will put that in my toolbox. 

Need to finish most of the other flags first.

**Comment by ChriMarMe** [comment]
When implementing the `--cache` option for routing tables of IPv6, I encountered legacy compatibility for really old kernel versions. There is a file opened and if that fails, it falls back to a the `-fib` function but with a filter passed to it. 
I asked the maintainer of net-tools what this is about. See: https://github.com/ecki/net-tools/issues/34
Here we have some additional informations about routing cache:
https://serverfault.com/questions/1091128/why-i-get-cache-in-the-output-of-ip-route-get
https://workshop.netfilter.org/2013/wiki/images/2/2a/DaveM_route_cache_removed_nfws2013.pdf

So my takeaway is that routing cache was removed for IPv4/6 some time ago already and the `--cache`-flag is dragged along for legacy reasons.

For now I will skip the implementation of `--cache`. Let me what you think @10000TB @andrewsun2898

**Comment by 10000TB** [comment]
@ChriMarMe @jensdrenhaus Hi, srry for the delay! -- re. socket types and address family. 

I mostly aggress with the assesment, except we would like TCP/UDP included too. I think that is beneficial to include. So the requirements from my end are as follows. 

Socket types:
* Required 
    * Raw
    * Unix
    * Tcp
    * Udp

* Not Required
    * ax25
    * sctp
    * Ipx
    * Netrom

Address Family
* Required 
    * IPv4
    * IPv6
    
* Not Required 
    * ax25 (Amateur radio AX.25 protocol)
    * netrom (AMPR NET/ROM)
    * ipx (Novell IPX)
    * ddp(Appletalk DDP) 
    * x25 (ITU-T X.25 / ISO-8208 protocol)

**Comment by 10000TB** [comment]
> When implementing the `--cache` option for routing tables of IPv6, I encountered legacy compatibility for really old kernel versions. There is a file opened and if that fails, it falls back to a the `-fib` function but with a filter passed to it. I asked the maintainer of net-tools what this is about. See: [ecki/net-tools#34](https://github.com/ecki/net-tools/issues/34) Here we have some additional informations about routing cache: https://serverfault.com/questions/1091128/why-i-get-cache-in-the-output-of-ip-route-get https://workshop.netfilter.org/2013/wiki/images/2/2a/DaveM_route_cache_removed_nfws2013.pdf
> 
> So my takeaway is that routing cache was removed for IPv4/6 some time ago already and the `--cache`-flag is dragged along for legacy reasons.
> 
> For now I will skip the implementation of `--cache`. Let me know what you think @10000TB @andrewsun2898

In that case, how do we display routing cache ? -- and how do we different route from cache versus new lookup need happen ?

I tried on my 6.6 kernel devbox, which has `netstat` 2.10



```
xuehaohu@worldpeace:~$ netstat --version
net-tools 2.10
Fred Baumgarten, Alan Cox, Bernd Eckenfels, Phil Blundell, Tuan Hoang, Brian Micek and others
+NEW_ADDRT +RTF_IRTT +RTF_REJECT +FW_MASQUERADE +I18N +SELINUX
AF: (inet) +UNIX +INET +INET6 +IPX +AX25 +NETROM +X25 +ATALK +ECONET +ROSE -BLUETOOTH
HW:  +ETHER +ARC +SLIP +PPP +TUNNEL -TR +AX25 +NETROM +X25 +FR +ROSE +ASH +SIT +FDDI +HIPPI +HDLC/LAPB +EUI64
xuehaohu@worldpeace:~$ uname -a
Linux 6.6.15-2rodete2-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.6.15-2rodete2 (2024-03-19) x86_64 GNU/Linux
xuehaohu@worldpeace:
```

```
xuehaohu@worldpeace:~$ netstat --help
...
...
        -C, --cache              display routing cache instead of FIB

xuehaohu@worldpeace:~$ netstat -6 -C
Active Internet connections (w/o servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State
tcp6       0      0 localhost:52868         localhost:9557          ESTABLISHED
tcp6       0      0 localhost:46904         localhost:9557          ESTABLISHED
tcp6       0      0 localhost:57110         localhost:9557          ESTABLISHED
tcp6       0      0 (redacted):49230     (redacted):5031 ESTABLISHED
tcp6       0      0 localhost:57150         localhost:9557          ESTABLISHED
tcp6       0      0 (redacted):35642     (redacted):5031 ESTABLISHED
tcp6       0      0 localhost:41834         localhost:9557          ESTABLISHED
tcp6       0      0 (redacted):58684     (redacted):5031    ESTABLISHED

```

**Comment by ChriMarMe** [comment]
> @ChriMarMe @jensdrenhaus Hi, srry for the delay! -- re. socket types and address family.
> 
> I mostly agree with the assesment put by @jensdrenhaus , except we would like TCP/UDP included too. I think that is beneficial to include. So the requirements from my end are as follows.
> 
> Socket types:
> 
>     * Required
>       
>       * Raw
>       * Unix
>       * Tcp
>       * Udp
> 
>     * Not Required
>       
>       * ax25
>       * sctp
>       * Ipx
>       * Netrom
> 
> 
> Address Family
> 
>     * Required
>       
>       * IPv4
>       * IPv6
> 
>     * Not Required
>       
>       * ax25 (Amateur radio AX.25 protocol)
>       * netrom (AMPR NET/ROM)
>       * ipx (Novell IPX)
>       * ddp(Appletalk DDP)
>       * x25 (ITU-T X.25 / ISO-8208 protocol)

Thanks for the clarification. I didnt mention TCP/UDP and IPv4/IPv6 because I assumed it as required already :) 



> > When implementing the `--cache` option for routing tables of IPv6, I encountered legacy compatibility for really old kernel versions. There is a file opened and if that fails, it falls back to a the `-fib` function but with a filter passed to it. I asked the maintainer of net-tools what this is about. See: [ecki/net-tools#34](https://github.com/ecki/net-tools/issues/34) Here we have some additional informations about routing cache: https://serverfault.com/questions/1091128/why-i-get-cache-in-the-output-of-ip-route-get https://workshop.netfilter.org/2013/wiki/images/2/2a/DaveM_route_cache_removed_nfws2013.pdf
> > So my takeaway is that routing cache was removed for IPv4/6 some time ago already and the `--cache`-flag is dragged along for legacy reasons.
> > For now I will skip the implementation of `--cache`. Let me know what you think @10000TB @andrewsun2898
> 
> In that case, how do we display routing cache ? -- and how do we different route from cache versus new lookup need happen ?
> 
> I tried on my 6.6 kernel devbox, which has `netstat` 2.10
> 
> ```
> xuehaohu@worldpeace:~$ netstat --version
> net-tools 2.10
> Fred Baumgarten, Alan Cox, Bernd Eckenfels, Phil Blundell, Tuan Hoang, Brian Micek and others
> +NEW_ADDRT +RTF_IRTT +RTF_REJECT +FW_MASQUERADE +I18N +SELINUX
> AF: (inet) +UNIX +INET +INET6 +IPX +AX25 +NETROM +X25 +ATALK +ECONET +ROSE -BLUETOOTH
> HW:  +ETHER +ARC +SLIP +PPP +TUNNEL -TR +AX25 +NETROM +X25 +FR +ROSE +ASH +SIT +FDDI +HIPPI +HDLC/LAPB +EUI64
> xuehaohu@worldpeace:~$ uname -a
> Linux 6.6.15-2rodete2-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.6.15-2rodete2 (2024-03-19) x86_64 GNU/Linux
> xuehaohu@worldpeace:
> ```
> 
> ```
> xuehaohu@worldpeace:~$ netstat --help
> ...
> ...
>         -C, --cache              display routing cache instead of FIB
> 
> xuehaohu@worldpeace:~$ netstat -6 -C
> Active Internet connections (w/o servers)
> Proto Recv-Q Send-Q Local Address           Foreign Address         State
> tcp6       0      0 localhost:52868         localhost:9557          ESTABLISHED
> tcp6       0      0 localhost:46904         localhost:9557          ESTABLISHED
> tcp6       0      0 localhost:57110         localhost:9557          ESTABLISHED
> tcp6       0      0 (redacted):49230     (redacted):5031 ESTABLISHED
> tcp6       0      0 localhost:57150         localhost:9557          ESTABLISHED
> tcp6       0      0 (redacted):35642     (redacted):5031 ESTABLISHED
> tcp6       0      0 localhost:41834         localhost:9557          ESTABLISHED
> tcp6       0      0 (redacted):58684     (redacted):5031    ESTABLISHED
> ```

Will have a look at it again. Prolly misunderstood some of the information.

**Comment by ChriMarMe** [comment]
> > When implementing the `--cache` option for routing tables of IPv6, I encountered legacy compatibility for really old kernel versions. There is a file opened and if that fails, it falls back to a the `-fib` function but with a filter passed to it. I asked the maintainer of net-tools what this is about. See: [ecki/net-tools#34](https://github.com/ecki/net-tools/issues/34) Here we have some additional informations about routing cache: https://serverfault.com/questions/1091128/why-i-get-cache-in-the-output-of-ip-route-get https://workshop.netfilter.org/2013/wiki/images/2/2a/DaveM_route_cache_removed_nfws2013.pdf
> > So my takeaway is that routing cache was removed for IPv4/6 some time ago already and the `--cache`-flag is dragged along for legacy reasons.
> > For now I will skip the implementation of `--cache`. Let me know what you think @10000TB @andrewsun2898
> 
> In that case, how do we display routing cache ? -- and how do we different route from cache versus new lookup need happen ?
> 
> I tried on my 6.6 kernel devbox, which has `netstat` 2.10
> 
> ```
> xuehaohu@worldpeace:~$ netstat --version
> net-tools 2.10
> Fred Baumgarten, Alan Cox, Bernd Eckenfels, Phil Blundell, Tuan Hoang, Brian Micek and others
> +NEW_ADDRT +RTF_IRTT +RTF_REJECT +FW_MASQUERADE +I18N +SELINUX
> AF: (inet) +UNIX +INET +INET6 +IPX +AX25 +NETROM +X25 +ATALK +ECONET +ROSE -BLUETOOTH
> HW:  +ETHER +ARC +SLIP +PPP +TUNNEL -TR +AX25 +NETROM +X25 +FR +ROSE +ASH +SIT +FDDI +HIPPI +HDLC/LAPB +EUI64
> xuehaohu@worldpeace:~$ uname -a
> Linux 6.6.15-2rodete2-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.6.15-2rodete2 (2024-03-19) x86_64 GNU/Linux
> xuehaohu@worldpeace:
> ```
> 
> ```
> xuehaohu@worldpeace:~$ netstat --help
> ...
> ...
>         -C, --cache              display routing cache instead of FIB
> 
> xuehaohu@worldpeace:~$ netstat -6 -C
> Active Internet connections (w/o servers)
> Proto Recv-Q Send-Q Local Address           Foreign Address         State
> tcp6       0      0 localhost:52868         localhost:9557          ESTABLISHED
> tcp6       0      0 localhost:46904         localhost:9557          ESTABLISHED
> tcp6       0      0 localhost:57110         localhost:9557          ESTABLISHED
> tcp6       0      0 (redacted):49230     (redacted):5031 ESTABLISHED
> tcp6       0      0 localhost:57150         localhost:9557          ESTABLISHED
> tcp6       0      0 (redacted):35642     (redacted):5031 ESTABLISHED
> tcp6       0      0 localhost:41834         localhost:9557          ESTABLISHED
> tcp6       0      0 (redacted):58684     (redacted):5031    ESTABLISHED
> ```

BTW: The output you have here is not route cache. It's IPv6 sockets listing.

**Comment by 10000TB** [comment]
> > > When implementing the `--cache` option for routing tables of IPv6, I encountered legacy compatibility for really old kernel versions. There is a file opened and if that fails, it falls back to a the `-fib` function but with a filter passed to it. I asked the maintainer of net-tools what this is about. See: [ecki/net-tools#34](https://github.com/ecki/net-tools/issues/34) Here we have some additional informations about routing cache: https://serverfault.com/questions/1091128/why-i-get-cache-in-the-output-of-ip-route-get https://workshop.netfilter.org/2013/wiki/images/2/2a/DaveM_route_cache_removed_nfws2013.pdf
> > > So my takeaway is that routing cache was removed for IPv4/6 some time ago already and the `--cache`-flag is dragged along for legacy reasons.
> > > For now I will skip the implementation of `--cache`. Let me know what you think @10000TB @andrewsun2898
> > 
> > 
> > In that case, how do we display routing cache ? -- and how do we different route from cache versus new lookup need happen ?
> > I tried on my 6.6 kernel devbox, which has `netstat` 2.10
> > ```
> > xuehaohu@worldpeace:~$ netstat --version
> > net-tools 2.10
> > Fred Baumgarten, Alan Cox, Bernd Eckenfels, Phil Blundell, Tuan Hoang, Brian Micek and others
> > +NEW_ADDRT +RTF_IRTT +RTF_REJECT +FW_MASQUERADE +I18N +SELINUX
> > AF: (inet) +UNIX +INET +INET6 +IPX +AX25 +NETROM +X25 +ATALK +ECONET +ROSE -BLUETOOTH
> > HW:  +ETHER +ARC +SLIP +PPP +TUNNEL -TR +AX25 +NETROM +X25 +FR +ROSE +ASH +SIT +FDDI +HIPPI +HDLC/LAPB +EUI64
> > xuehaohu@worldpeace:~$ uname -a
> > Linux 6.6.15-2rodete2-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.6.15-2rodete2 (2024-03-19) x86_64 GNU/Linux
> > xuehaohu@worldpeace:
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > xuehaohu@worldpeace:~$ netstat --help
> > ...
> > ...
> >         -C, --cache              display routing cache instead of FIB
> > 
> > xuehaohu@worldpeace:~$ netstat -6 -C
> > Active Internet connections (w/o servers)
> > Proto Recv-Q Send-Q Local Address           Foreign Address         State
> > tcp6       0      0 localhost:52868         localhost:9557          ESTABLISHED
> > tcp6       0      0 localhost:46904         localhost:9557          ESTABLISHED
> > tcp6       0      0 localhost:57110         localhost:9557          ESTABLISHED
> > tcp6       0      0 (redacted):49230     (redacted):5031 ESTABLISHED
> > tcp6       0      0 localhost:57150         localhost:9557          ESTABLISHED
> > tcp6       0      0 (redacted):35642     (redacted):5031 ESTABLISHED
> > tcp6       0      0 localhost:41834         localhost:9557          ESTABLISHED
> > tcp6       0      0 (redacted):58684     (redacted):5031    ESTABLISHED
> > ```
> 
> BTW: The output you have here is not route cache. It's IPv6 sockets listing.

okay. BTW, I just found out that `ss` is replacing `netstat`, as a more modern alternative. Could we double cache stuff in the sources of `ss` ? 

```
xuehaohu@worldpeace:~$ ss --help
Usage: ss [ OPTIONS ]
       ss [ OPTIONS ] [ FILTER ]
   -h, --help          this message
   -V, --version       output version information
   -n, --numeric       don't resolve service names
   -r, --resolve       resolve host names
   -a, --all           display all sockets
   -l, --listening     display listening sockets
   -o, --options       show timer information
   -e, --extended      show detailed socket information
   -m, --memory        show socket memory usage
   -p, --processes     show process using socket
   -T, --threads       show thread using socket
   -i, --info          show internal TCP information
       --tipcinfo      show internal tipc socket information
   -s, --summary       show socket usage summary
       --tos           show tos and priority information
       --cgroup        show cgroup information
   -b, --bpf           show bpf filter socket information
   -E, --events        continually display sockets as they are destroyed
   -Z, --context       display task SELinux security contexts
   -z, --contexts      display task and socket SELinux security contexts
   -N, --net           switch to the specified network namespace name

   -4, --ipv4          display only IP version 4 sockets
   -6, --ipv6          display only IP version 6 sockets
   -0, --packet        display PACKET sockets
   -t, --tcp           display only TCP sockets
   -M, --mptcp         display only MPTCP sockets
   -S, --sctp          display only SCTP sockets
   -u, --udp           display only UDP sockets
   -d, --dccp          display only DCCP sockets
   -w, --raw           display only RAW sockets
   -x, --unix          display only Unix domain sockets
       --tipc          display only TIPC sockets
       --vsock         display only vsock sockets
       --xdp           display only XDP sockets
   -f, --family=FAMILY display sockets of type FAMILY
       FAMILY := {inet|inet6|link|unix|netlink|vsock|tipc|xdp|help}

   -K, --kill          forcibly close sockets, display what was closed
   -H, --no-header     Suppress header line
   -O, --oneline       socket's data printed on a single line
       --inet-sockopt  show various inet socket options

   -A, --query=QUERY, --socket=QUERY
       QUERY := {all|inet|tcp|mptcp|udp|raw|unix|unix_dgram|unix_stream|unix_seqpacket|packet|packet_raw|packet_dgram|netlink|dccp|sctp|vsock_stream|vsock_dgram|tipc|xdp}[,QUERY]

   -D, --diag=FILE     Dump raw information about TCP sockets to FILE
   -F, --filter=FILE   read filter information from FILE
       FILTER := [ state STATE-FILTER ] [ EXPRESSION ]
       STATE-FILTER := {all|connected|synchronized|bucket|big|TCP-STATES}
         TCP-STATES := {established|syn-sent|syn-recv|fin-wait-{1,2}|time-wait|closed|close-wait|last-ack|listening|closing}
          connected := {established|syn-sent|syn-recv|fin-wait-{1,2}|time-wait|close-wait|last-ack|closing}
       synchronized := {established|syn-recv|fin-wait-{1,2}|time-wait|close-wait|last-ack|closing}
             bucket := {syn-recv|time-wait}
                big := {established|syn-sent|fin-wait-{1,2}|closed|close-wait|last-ack|listening|closing}
xuehaohu@worldpeace:~$
xuehaohu@worldpeace:~$
xuehaohu@worldpeace:~$
xuehaohu@worldpeace:~$ man ss

SS(8)                                          System Manager's Manual                                         SS(8)

NAME
       ss - another utility to investigate sockets

SYNOPSIS
       ss [options] [ FILTER ]

DESCRIPTION
       ss  is used to dump socket statistics. It allows showing information similar to netstat.  It can display more
       TCP and state information than other tools.

OPTIONS
       When no option is used ss displays a list of open non-listening sockets (e.g. TCP/UNIX/UDP) that have  estab‐
       lished connection.
...
...
SEE ALSO
       ip(8),
       RFC 793 - https://tools.ietf.org/rfc/rfc793.txt (TCP states)

AUTHOR
       ss was written by Alexey Kuznetsov, <kuznet@ms2.inr.ac.ru>.

       This manual page was written by Michael Prokop <mika@grml.org> for the Debian project (but  may  be  used  by
       others).


```

**Comment by ChriMarMe** [comment]
@10000TB 
>BTW, I just found out that ss is replacing netstat, as a more modern alternative. Could we double cache stuff in the sources of ss?

What do you mean by "double cache"?

**Comment by ChriMarMe** [comment]
Restructures the commits. I know they're still quite big each. Every commit implements the required functionality for the specific main capability of netstat.

**Comment by 10000TB** [comment]
> Could we double cache stuff in the sources of `ss` ?

my bad...typo. Meant to ask to you to double **check** cache stuff in `ss`.

**Comment by ChriMarMe** [comment]
> > Could we double cache stuff in the sources of `ss` ?
> 
> my bad...typo. Meant to ask to you to double **check** cache stuff in `ss`.

`ss` does not support route listing at all. 

iptables2 code gives hints about route cache being removed.

Also the manpage of iptables2-package/ip-route subcommand (the ip command) gives a hint to that:
[ip-route.8 man-page](https://github.com/iproute2/iproute2/blob/39e4b6f5f315680ac914187b1d5526c005b8c0d8/man/man8/ip-route.8.in#L1356C10-L1357C75)
```
.SH NOTES
Starting with Linux kernel version 3.6, there is no routing cache for IPv4
anymore. Hence
.B "ip route show cached"
will never print any entries on systems with this or newer kernel versions.
```

For IPv6 I found [this](https://vincent.bernat.ch/en/blog/2017-ipv6-route-lookup-linux) source which states, that route cache is still in the kernel but only creates a cache entry if a PMTU (Path MTU Discovery) exception happens. 

> Starting from Linux 4.2 [commit 45e4fd26683c](https://git.kernel.org/pub/scm/linux/kernel/git/davem/net-next.git/commit/?id=45e4fd26683c9a5f88600d91b08a484f7f09226a), only a PMTU exception would create a cache entry. 
> A router doesn’t have to handle these exceptions, so only hosts would get cache entries.

> And they should be pretty rare. Martin KaFai Lau, from Facebook, [explains](https://www.spinics.net/lists/netdev/msg330403.html):
> Out of all IPv6 RTF_CACHE routes that are created, the percentage that has a different MTU is very small. 
> In one of our end-user facing proxy server, only 1k out of 80k RTF_CACHE routes have a smaller MTU. 
> For our DC traffic, there is no MTU exception.

The implementation of `--cache` does not bring any benefit in my opinion.

**Comment by 10000TB** [comment]
Okay. Thank you Chris for summarizing the situation here.

I think this is taking a while to discuss. I think I am now ok with skipping --cache for ipv4 for netstat.

**Review by rminnich — COMMENTED** [review]
The pattern we have started to use is to define a struct for all the globals, and set up the flags and such in a run() function.

This lets us remove global variables, which will make it possible to use this command in bare metal environments. I am hoping you will consider it.

File: cmds/core/netstat/netstat_test.go
Comment (rminnich) [inline cmds/core/netstat/netstat_test.go:76]: if you adapt the idea of using a struct, you won't need resetFlags.

File: pkg/netstat/format.go
Comment (rminnich) [inline pkg/netstat/format.go:182]: and those are dead. Good call to make it constant

File: cmds/core/netstat/netstat_linux.go
Comment (rminnich) [inline cmds/core/netstat/netstat_linux.go:79]: why is there a unixFlag? 
Can't you just check runtime.GOOS? If you do, then when it is built, nice things happen. 
e.g. 
if runtime.GOOS == "linux" {
}
that code will not appear in the binary on non-Linux systems.

**Review by rminnich — COMMENTED** [review]
I hate to be picky, but the way this is written, it's a lot of code needing tests. 

A friend at meta has had good luck with asking chatgpt to write tests for Go.

File: cmds/core/netstat/netstat_linux.go
Comment (ChriMarMe) [reply cmds/core/netstat/netstat_linux.go:79]: unix flag triggers printing of unix sockets instead of IP sockets

File: cmds/core/netstat/netstat_test.go
Comment (ChriMarMe) [reply cmds/core/netstat/netstat_test.go:76]: true. Already working on the struct.

**Comment by ChriMarMe** [comment]
> I hate to be picky, but the way this is written, it's a lot of code needing tests.
> 
> A friend at meta has had good luck with asking chatgpt to write tests for Go.

Being picky is what helps me being a better dev. So thank you :)

File: cmds/core/netstat/netstat_linux.go
Comment (10000TB) [reply cmds/core/netstat/netstat_linux.go:51]: oh, I see you have more readable info about those flags in flags definition down below in `main`. Maybe simply leave a note saying check down below in main for more info

File: cmds/core/netstat/netstat_linux.go
Comment (ChriMarMe) [reply cmds/core/netstat/netstat_linux.go:58]: Type indeed

File: cmds/core/netstat/netstat_linux.go
Comment (ChriMarMe) [reply cmds/core/netstat/netstat_linux.go:20]: Agreed

File: cmds/core/netstat/netstat_linux.go
Comment (ChriMarMe) [reply cmds/core/netstat/netstat_linux.go:97]: Will do.

File: cmds/core/netstat/netstat_linux.go
Comment (ChriMarMe) [reply cmds/core/netstat/netstat_linux.go:51]: Moved the main function up right below the Flags struct and wrote a comment to the Flags struct.

File: cmds/core/netstat/netstat_linux.go
Comment (ChriMarMe) [reply cmds/core/netstat/netstat_linux.go]: done

**Comment by ChriMarMe** [comment]
Do I miss anything? Anything I can do to speed up the process?

**Review by rminnich — APPROVED** [review]
let's go. Maybe this is incomplete, but that can be fixed.

**Review by 10000TB — APPROVED** [review]
Discussed offline, we will go with this version for now.

If we need squash this, after discussion, we prefer to keep commit hisotry in this PR to show the trajectory, otherwise, this PR is too large to understand.

Next steps,

Andrew / David integrate it, try it and share feedback on if there need anything else from Chris

**Pull request merged**
